src/typechecker.erl

```diff
@@ -20,6 +20,36 @@
 
 -type type() :: erl_parse:abstract_type().
 
+-type type_error() ::
+    #{error_type := {undef, call | record | user_type | remote_type | variable}
+                    | unexported_remote_type | var | char | atom
+                    | string | int | float | improper_tail | cons_pat
+                    | cons | nil | argument_length_mismatch | call_args
+                    | {call, args | arity | context} | mfa
+                    | return | expected_fun_type | op_type_too_precise
+                    | {arith_op, arguments | expression}
+                    | {int_op, arguments | expression}
+                    | non_number_argument_to_plus | non_number_argument_to_minus
+                    | non_boolean_argument_to_not
+                    | unary_op | non_integer_argument_to_bnot
+                    | logic_op | {rel_op, arguments | expression}
+                    | {list_op, arguments | expression}
+                    | operator_pattern | pattern | tuple
+                    | {bit, arguments | expression} | generator | bin_generator
+                    | check_clauses | record | record_pattern | record_update
+                    | {comprehension, bit_string | list | binary}
+                    | lambda | cyclic_type_vars,
+      line           => erl_anno:anno(),
+      module         => atom(),
+      name           => atom(),
```

Comment:
I would suggest we call this field 'function' or something along those lines which is a little more descriptive than 'name'.

Reply:
In that case I should also change
```Erlang
handle_type_error({type_error, tyVar, LINE, Var, VarTy, Ty}) ->
```
to return
```
    #{error_type    => var,
      line          => LINE,
      expression    => Var,
      actual_type   => VarTy,
      expected_type => Ty
      };
```

Reply:
Sorry, I don't follow. Can you elaborate?

Reply:
At the moment it uses the `name` field which if renamed would not make sense there.
I don't want a field for just this case so moving `Var` to `expression` is my preferred solution.

Reply:
Right. So you had chosen 'name' specifically because it's ambiguous and can be used in many different context. 
Yeah, using 'expression' for variables sounds like a good solution.

Reply:
What about `{undef, record}`, `{undef, Type}`, `unexported_remote_type`?
Perhaps the whole MFA should be in `expression`? I avoided doing so since I don't know what the type of most things in `expression` is.

Reply:
I agree about encoding stuff in the other fields so the error_type can just be `undefined` or `unexported`, but in these cases we don't have an expression. These errors are are coming from a bad type in a spec or a type definition. The AST types are defined here: https://github.com/erlang/otp/blob/master/lib/stdlib/src/erl_parse.yrl#L605

Reply:
I don't know. Maybe put the bad record or type in a `type` field? Or `bad_type`...